cmd/zrepl/zrepl.c

```diff
@@ -0,0 +1,976 @@
+	if (rc == -1) {
+		ZREPL_ERRLOG("Unable to get IP"
+		    " with err:%d\n", errno);
+		goto exit;
+	}
+
+	packet = kmem_alloc((sizeof (mgmt_ack_t) + sizeof (*hdr)) *
+	    sizeof (char), KM_SLEEP);
+	bcopy(hdr, packet, sizeof (*hdr));
+	p = packet + sizeof (*hdr);
+	bcopy(&mgmt_ack, p, sizeof (mgmt_ack));
+	count = write(sfd, packet, (sizeof (*hdr) + sizeof (mgmt_ack_t)));
+	if (count == -1) {
+		ZREPL_ERRLOG("Write to socket failed"
+		    " with err:%d\n", errno);
+		rc = -1;
+		goto exit;
+	}
+exit:
+	uzfs_zinfo_drop_refcnt(zinfo, false);
+	return (rc);
+}
+
+/*
+ * TODO: IMHO, this should be a through away API.
+ * Side Car has to find a more smart way to pass
+ * ISCSI Controller IP address.
+ */
+static char *
+get_controller_ip_address(void)
+{
+	size_t nbytes;
+	char *buf = NULL;
+
+	FILE *fp = fopen("/var/openebs/controllers.conf", "r");
+	if (fp == NULL) {
+		printf("Error opening file\n");
+		return (buf);
+	}
+
+	buf = kmem_alloc(1024, KM_SLEEP);
+	nbytes = fread(buf, sizeof (char), 1024, fp);
+
+	if (nbytes == 0) {
+		printf("Read error\n");
+		return (buf);
+	}
+	printf("buffer value is %s", buf);
+	return (buf);
+}
+
+/*
+ * One thread per replica, which will be
+ * responsible for initial handshake and
+ * exchanging info like IP add, port etc.
+ */
+static void
+uzfs_zvol_mgmt_thread(void *arg)
+{
+
+	int sfd, efd, rc, count;
+	struct epoll_event event;
+	struct sockaddr_in istgt_addr;
+	zvol_io_hdr_t hdr = {0, };
+	struct epoll_event *events = NULL;
+	char *name = NULL;
+	char *buf = NULL;
+
+
+	sfd = create_and_bind(mgmt_port);
+	if (sfd == -1) {
+		goto exit;
+	}
+
+	rc = make_socket_non_blocking(sfd);
+	if (rc == -1) {
+		goto exit;
+	}
+
+	buf = get_controller_ip_address();
+	if (buf == NULL) {
+		printf("parsing IP address did not work\n");
+		goto exit;
+	}
+	printf("Controller IP address is: %s", buf);
+	bzero((char *)&istgt_addr, sizeof (istgt_addr));
+	istgt_addr.sin_family = AF_INET;
+	istgt_addr.sin_addr.s_addr = inet_addr(buf);
+	istgt_addr.sin_port = htons(6060);
+	free(buf);
+retry:
+	rc = connect(sfd, (struct sockaddr *)&istgt_addr, sizeof (istgt_addr));
+	if ((rc == -1) && (errno == EINTR)) {
+		ZREPL_ERRLOG("Failed to connect to istgt_controller"
+		    " with err:%d\n", errno);
+		sleep(10);
+		goto retry;
+	} else {
+		printf("Connection to TGT controller successful\n");
+		ZREPL_LOG("Connection to TGT controller iss successful\n");
+	}
+
+	efd = epoll_create1(0);
+	if (efd == -1) {
+		ZREPL_ERRLOG("epoll_create() failed with errno:%d\n", errno);
+		goto exit;
+	}
+
+	event.data.fd = sfd;
+	event.events = EPOLLIN | EPOLLET | EPOLLERR |
+	    EPOLLHUP | EPOLLRDHUP;
+	rc = epoll_ctl(efd, EPOLL_CTL_ADD, sfd, &event);
+	if (rc == -1) {
+		ZREPL_ERRLOG("epoll_ctl() failed with errno:%d\n", errno);
+		goto exit;
+	}
+
+	/* Buffer where events are returned */
+	events = calloc(MAXEVENTS, sizeof (event));
+
+	/* The event loop */
+	while (1) {
+		int n, i;
+		n = epoll_wait(efd, events, MAXEVENTS, -1);
```

Comment:
why do we bother with epoll if all we want to do is read data from one file descriptor?

Reply:
Yes. I changed it. It was very hard for me to address this comment.

Reply:
I hope I did not provide a misleading advice in this case :-)

Reply:
No, absolutely not. It was constructive comment.